frontends/common/constantFolding.cpp

```diff
@@ -237,6 +237,36 @@ const IR::Node *DoConstantFolding::preorder(IR::ArrayIndex *e) {
     return e;
 }
 
+// Returns true if the given expression is of the form "some_table.apply().action_run."
+bool isActionRun(const IR::Expression *e, const ReferenceMap *refMap) {
+    const auto *actionRunMem = e->to<IR::Member>();
+    if (!actionRunMem) return false;
+    if (actionRunMem->member.name != IR::Type_Table::action_run) return false;
+
+    const auto *applyMce = actionRunMem->expr->to<IR::MethodCallExpression>();
+    if (!applyMce) return false;
+    const auto *applyMceMem = applyMce->method->to<IR::Member>();
+    if (!applyMceMem) return false;
+    if (applyMceMem->member.name != IR::IApply::applyMethodName) return false;
+
+    const auto *tablePathExpr = applyMceMem->expr->to<IR::PathExpression>();
+    if (!tablePathExpr) return false;
+    const auto *tableDecl = refMap->getDeclaration(tablePathExpr->path);
+    if (!tableDecl) return false;
+
+    return tableDecl->is<IR::P4Table>();
+}
+
+const IR::Node *DoConstantFolding::preorder(IR::SwitchCase *c) {
+    // Action enum switch case labels must be action names.
+    // We only want to fold the label expression after it is inspected by the typechecker.
+    const auto *parent = static_cast<const IR::SwitchStatement *>(getContext()->node);
+    if (typesKnown || !isActionRun(parent->expression, refMap)) visit(c->label);
+    visit(c->statement);
+    prune();
```

Comment:
```suggestion
    if (isActionRun(parent->expression, refMap)) prune()
```

... this way we can just defer to the standard way visitor does visits if it is not in `action_run`.

Reply:
Seems reasonable to me. But the equivalent would be:
```
diff --git a/frontends/common/constantFolding.cpp b/frontends/common/constantFolding.cpp
index ec97799bd..83a8ecd92 100644
--- a/frontends/common/constantFolding.cpp
+++ b/frontends/common/constantFolding.cpp
@@ -259,11 +259,13 @@ bool isActionRun(const IR::Expression *e, const ReferenceMap *refMap) {

 const IR::Node *DoConstantFolding::preorder(IR::SwitchCase *c) {
     // Action enum switch case labels must be action names.
-    // We only want to fold the label expression after it is inspected by the typechecker.
+    // Do not fold the switch case's 'label' expression so that it can be inspected by the
+    // TypeInference pass.
     const auto *parent = static_cast<const IR::SwitchStatement *>(getContext()->node);
-    if (typesKnown || !isActionRun(parent->expression, refMap)) visit(c->label);
-    visit(c->statement);
-    prune();
+    if (isActionRun(parent->expression, refMap)) {
+        visit(c->statement);
+        prune();
+    }
     return c;
 }
```

Reply:
oh, sorry, of course, the statement still needs to be visited.